src/gleam/bit_array.gleam

```diff
@@ -97,6 +97,154 @@ pub fn to_string(bits: BitArray) -> Result(String, Nil) {
 @external(erlang, "gleam_stdlib", "identity")
 fn unsafe_to_string(a: BitArray) -> String
 
+/// Converts a bit array to a string. Invalid bits are passed to the provided
+/// callback and its result is included in the final string in place of the
+/// invalid data.
+///
+/// ## Examples
+///
+/// ```gleam
+/// to_string_lossy(<<"A":utf8, 0x80, "1":utf8, 0:size(5)>>, fn(_) { "�" })
+/// // -> "A�1�"
+/// ```
+/// 
+pub fn to_string_lossy(
+  bits: BitArray,
+  map_invalid_bits: fn(BitArray) -> String,
+) -> String {
+  to_string_lossy_impl(bits, map_invalid_bits, "")
+}
+
+@target(erlang)
+fn to_string_lossy_impl(
+  bits: BitArray,
+  map_invalid_bits: fn(BitArray) -> String,
+  acc: String,
+) -> String {
+  case bits {
+    <<>> -> acc
+
+    <<x:utf8_codepoint, rest:bits>> ->
+      to_string_lossy_impl(
+        rest,
+        map_invalid_bits,
+        acc <> string.from_utf_codepoints([x]),
+      )
+
+    <<x:bytes-1, rest:bits>> ->
+      to_string_lossy_impl(rest, map_invalid_bits, acc <> map_invalid_bits(x))
+
+    _ -> acc <> map_invalid_bits(bits)
+  }
+}
+
+// The following is the same as the above function but supports the JavaScript
+// target due to not using the `utf8_codepoint` bit array segment type. Once
+// the JavaScript target supports `utf8_codepoint` this function should be
+// removed.
+@target(javascript)
```

Comment:
It's a shame so much is needed for JavaScript, and this introduces a new `@target`, which we have been working hard to remove.

How challenging would it be to have `utf8_codepoint` support in JavaScript? That seems like it would be a better solution.

Reply:
Bit array pattern matching in JS would need to be expanded to support dynamic sizes first, then `utf_codepoint` could be implemented on top of that.

I think it's reasonable to wait for that to happen, or alternatively a `@target` can be avoided by either using the longer JS-compatible version on both targets (it works fine on Erlang), or by implementing the JS version directly in JS as an external function.